lib/std/smartptrs.nim

```diff
@@ -0,0 +1,130 @@
+#
+#
+#            Nim's Runtime Library
+#        (c) Copyright 2021 Nim contributors
+#
+#    See the file "copying.txt", included in this
+#    distribution, for details about the copyright.
+
+## C++11 like smart pointers. They always use the shared allocator.
```

Comment:
this needs to work with `--threads:off` otherwise it'll force client code to use `--threads:on` just so they can use some library that uses smartptrs, which can negatively impact performance

Reply:
```suggestion
## C++11 like smart pointers. They always use the shared allocator.
## Experimental API, subject to change.
```

the disclaimer is needed, it's too early to ciment this API; either that or add back `-d:nimExperimentalSmartptrs` which allows code to opt-in to it and get more real life usage to iron out API instead of cimenting it from day 1

Reply:
> @timotheecour consider using the "Start Review" tool in GH so that we're not getting notifications for every new comment you add.

That was never an issue for me. Are there email clients that don't support threads? Why would people not use grouping by thread?

Reply:
See here we disagree, as I am using smartptrs and would like this api formalized. I think thats the whole point of submitting it in stdlib. Why was it in fusion before that? So that people can use it, and suggest improvements. Thats why I protested on every change that came out of the blue, like destructor field and the `get` procs.

Reply:
will put:

```nim
when not compileOption("threads"):
  {.error: "smartptrs requires --threads:on option.".}
```

instead

Reply:
the goal is to get the api formalized eventually, but what makes you think you'll come up with a design that won't need to be broken in some backward incompatible way, once more usage is gathered? fusion didn't garner as much visibility as having it in stdlib. With this disclaimer, you set appropriate expectations. Many stdlib modules submitted had to go through a few iterations (read: separate PRs) to get the API right.

Reply:
Alright, disclaimer is fine with me.

Reply:
It also depends, if people could have given examples, I could have come up with something that works. Till that didn't happen I need to approximate their usage scenarios.

Reply:
i gave a concrete example with importcpp types that I'd expect to work in std/smartptrs

I have other examples in mind, eg a std::vector<T> - like type re-implemented in nim (ie, not using seq but using instead realloc, capacity, size), similar to the C++ implementation of `std::vector<T>`.

Reply:
I use locks (and other pthread primitives) with smartptrs and they work. Its not unreasonable for me to think your stuff would work as well. however if you could try to make an actual example, compile it and see what errors it gives, we can take a look.